src/java.base/share/classes/java/security/PEMRecord.java

```diff
@@ -0,0 +1,150 @@
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package java.security;
+
+import jdk.internal.javac.PreviewFeature;
+
+import sun.security.util.Pem;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+
+/**
+ * {@code PEMRecord} is a {@link DEREncodable} that represents Privacy-Enhanced
+ * Mail (PEM) data by its type and Base64 form.  {@link PEMDecoder} and
+ * {@link PEMEncoder} use {@code PEMRecord} when representing the data as a
+ * Java API cryptographic object is not desired or there is no other
+ * {@code DEREncodable} for the type.
+ *
+ * <p>{@code PEMRecord} may have a null {@code type} and {@code pem} when
+ * there is no PEM data.  If there is PEM data, {@code type} and
+ * {@code pem} will both be non-null. {@code leadingData} may be null if the
+ * decoded data only contains PEM data. All values can never be null.
+ *
+ * <p> During the instantiation of this record, there is no validation for the
+ * {@code type} or {@code pem}.
+ *
+ * <p>This class is thread-safe.  {@code leadingData} is not
+ * defensively copied.
+ *
+ * @param type The type identifier in the PEM header.  For a public key,
+ * {@code type} would be "PUBLIC KEY".
+ * @param pem Any data between the PEM header and footer.
+ * @param leadingData Any non-PEM data read during the decoding process
+ * before the PEM header. This can be useful when reading metadata that
+ * accompanies PEM data.
+ *
+ * @spec https://www.rfc-editor.org/info/rfc7468
+ *       RFC 7468: Textual Encodings of PKIX, PKCS, and CMS Structures
+ *
+ * @see PEMDecoder
+ * @see PEMEncoder
+ */
+@PreviewFeature(feature = PreviewFeature.Feature.PEM_API)
+public record PEMRecord(String type, String pem, byte[] leadingData)
+    implements DEREncodable {
+
+    /**
+     * Creates a {@code PEMRecord} instance with the given parameters.
+     *
+     * @param type the type identifier in the PEM header and footer, or
+     *             {@code null} if there is no PEM data.
+     * @param pem the data between the PEM header and footer.
+     * @param leadingData Any non-PEM data read during the decoding process
+     *                    before the PEM header.  This value maybe {@code null}.
+     * @throws IllegalArgumentException on incorrect input values
+     */
+    public PEMRecord(String type, String pem, byte[] leadingData) {
+        if (type == null && pem == null && leadingData == null) {
+            throw new IllegalArgumentException("All values may not be null.");
+        }
+
+        if (type == null && pem != null || type != null && pem == null) {
+            throw new IllegalArgumentException("\"type\" and \"pem\" must be" +
+                " both null or non-null");
```

Comment:
If type and pem can be both null, and leadingData is also null (which is allowed per spec) then that should not be an error?

This check may need to be updated now that non-PEM data is an error.

Reply:
code has been changed